File: ts/src/alias.ts
/* Copyright (c) 2021-2026 Richard Rodger, MIT License */


import type { Val } from './type'

import { cmpCodePoint } from './keyorder'
import { spreadSnapKey } from './val/MapVal'


function expandAliases(root: Val, snapmap: Map<string, Val>): void {
  if (true !== (root as any).isMap) {
    return
  }

  const seen = new Set<Val>()

  const visit = (v: any, stack: string[]): void => {
    if (null == v || true !== v.isVal || seen.has(v)) {
      return
    }
    seen.add(v)

    if (true === v.isRef) {
      const name: string | undefined = v.aliasName
      if (undefined === name) {
        return
      }
      v.expansion = undefined
      if (stack.includes(name)) {
        return
      }
      const target: Val | undefined =
        snapmap.get(spreadSnapKey(v)) ?? (root as any).peg[name]
      if (null == target) {
        return
      }
      v.expansion = target
      visit(target, [...stack, name])
      return
    }

    if (true === v.isMap) {
      // A declaration is reached through its references, each under
      // its own name, never as a child: a self-reference inside it
      // is a knot only from inside.
      const keys = Object.keys(v.peg)
        .filter((k: string) => !v.aliasKeys.includes(k))
        .sort(cmpCodePoint)
      for (const k of keys) {
        visit(v.peg[k], stack)
      }
    }
    else if (Array.isArray(v.peg)) {
      for (const e of v.peg) {
        visit(e, stack)
      }
    }
    else if (null != v.peg && true === v.peg.isVal) {
      visit(v.peg, stack)
    }

    if ((true === v.isMap || true === v.isList) && null != v.spread.cj) {
      visit(v.spread.cj, stack)
    }
  }

  visit(root, [])
} /* node:coverage ignore next 5 */


export {
  expandAliases,
}

File: go/alias.go
/* Copyright (c) 2026 Richard Rodger, MIT License */

package aontu

import "sort"

func expandAliases(root Val, snapmap map[string]Val) {
	rm, ok := root.(*MapVal)
	if !ok {
		return
	}

	seen := map[Val]bool{}

	var visit func(v Val, stack []string)
	visit = func(v Val, stack []string) {
		if nil == v || seen[v] {
			return
		}
		seen[v] = true

		switch n := v.(type) {
		case *RefVal:
			name, isAlias := n.aliasName()
			if !isAlias {
				return
			}
			n.expansion = nil
			for _, s := range stack {
				if s == name {
					return
				}
			}
			target, snapped := snapmap[refSnapKey(n)]
			if !snapped {
				target = rm.peg[name]
			}
			if nil == target {
				return
			}
			n.expansion = target
			visit(target, append(append([]string{}, stack...), name))

		case *MapVal:
			// A declaration is reached through its references, each
			// under its own name, never as a child: a self-reference
			// inside it is a knot only from inside.
			keys := make([]string, 0, len(n.keys))
			for _, k := range n.keys {
				if !n.isAliasKey(k) {
					keys = append(keys, k)
				}
			}
			sort.Strings(keys)
			for _, k := range keys {
				visit(n.peg[k], stack)
			}
			if nil != n.spread {
				visit(n.spread, stack)
			}

		case *ListVal:
			for _, e := range n.peg {
				visit(e, stack)
			}
			if nil != n.spread {
				visit(n.spread, stack)
			}

		case *ConjunctVal:
			for _, e := range n.peg {
				visit(e, stack)
			}

		case *DisjunctVal:
			for _, e := range n.peg {
				visit(e, stack)
			}

		case *FuncVal:
			for _, e := range n.peg {
				visit(e, stack)
			}

		case *PlusOpVal:
			for _, e := range n.peg {
				visit(e, stack)
			}

		case *PrefVal:
			visit(n.peg, stack)

		case *ExpectVal:
			visit(n.peg, stack)
		}
	}

	visit(root, nil)
}
